akka-persistence-typed-tests/src/test/scala/akka/persistence/typed/ReplicatedEventSourcingTaggingSpec.scala

```diff
@@ -0,0 +1,111 @@
+        replicationContext =>
+          EventSourcedBehavior[Command, String, State](
+            replicationContext.persistenceId,
+            State(Set.empty),
+            (state, command) =>
+              command match {
+                case Add(string, ack) =>
+                  if (state.strings.contains(string)) Effect.none.thenRun(_ => ack ! Done)
+                  else Effect.persist(string).thenRun(_ => ack ! Done)
+                case GetStrings(replyTo) =>
+                  replyTo ! state.strings
+                  Effect.none
+              },
+            (state, event) => state.copy(strings = state.strings + event))
+          // use withTagger to define tagging logic
+            .withTagger(
+              event =>
+                // don't apply tags if event was replicated here, it already will appear in queries by tag
+                // as the origin replica would have tagged it already
+                if (replicationContext.origin != replicationContext.replicaId) Set.empty
+                else if (event.length > 10) Set("long-strings", "strings")
+                else Set("strings")))
+      // #simple-tag
+    }
+  }
+
+}
+
+class ReplicatedEventSourcingTaggingSpec
+    extends ScalaTestWithActorTestKit(PersistenceTestKitPlugin.config)
+    with AnyWordSpecLike
+    with LogCapturing
+    with Eventually {
+  import ReplicatedEventSourcingTaggingSpec._
+  val ids = new AtomicInteger(0)
+  def nextEntityId = s"e-${ids.getAndIncrement()}"
+  "ReplicatedEventSourcing" should {
+    "allow for tagging of events using the replication context" in {
+      val entityId = nextEntityId
+      val probe = createTestProbe[Done]()
+      val r1 = spawn(ReplicatedStringSet(entityId, ReplicaId1, AllReplicas))
+      val r2 = spawn(ReplicatedStringSet(entityId, ReplicaId2, AllReplicas))
+      r1 ! ReplicatedStringSet.Add("from r1", probe.ref)
+      r2 ! ReplicatedStringSet.Add("from r2", probe.ref)
+      probe.receiveMessages(2)
+      r1 ! ReplicatedStringSet.Add("a very long string from r1", probe.ref)
+      probe.receiveMessages(1)
+
+      val allEvents = Set("from r1", "from r2", "a very long string from r1")
+      for (replica <- r1 :: r2 :: Nil) {
+        eventually {
+          val probe = testKit.createTestProbe[Set[String]]()
+          replica ! ReplicatedStringSet.GetStrings(probe.ref)
+          probe.receiveMessage() should ===(allEvents)
+        }
+      }
+
+      /* FIXME can't verify tags with testkit ;(
+      val query =
+        PersistenceQuery(system).readJournalFor[CurrentEventsByTagQuery](PersistenceTestKitReadJournal.Identifier)
+
+      val stringTaggedEvents = query.currentEventsByTag("strings", NoOffset).runWith(Sink.seq).futureValue
+      stringTaggedEvents.map(_.event).toSet should equal(allEvents)
+
+      val longStrings = query.currentEventsByTag("long-strings", NoOffset).runWith(Sink.seq).futureValue
+      longStrings should have size (1)
+     */
```

Comment:
need to actually verify tags were written here, so I guess I need to use some other journal than the testkit or add tagging support to the testkit.